src/main/java/Duke.java

```diff
@@ -1,10 +1,45 @@
+import command.Command;
+import exception.DukeException;
+import storage.Storage;
+import task.*;
+import ui.Ui;
+
 public class Duke {
+    private TaskList tasks;
+    private Ui ui;
+    private Storage storage;
+
+    public Duke() {
+        storage = new Storage("data/duke.txt");
+        ui = new Ui();
+        try {
+            tasks = new TaskList(storage);
+        } catch (DukeException e) {
+            ui.printErrorMsg(e);
+        }
+    }
+
+    /**
+     * Run loop of Duke.
+     */
+    public void run() {
+        boolean running = true;
+        ui.printHelloMsg();
+        while (running) {
+            try {
+                String fullCommand = ui.readCommand();
+                Command c = Parser.parseInput(fullCommand);
+                c.execute(tasks, ui, storage);
+                running = !c.isExit();
```

Comment:
this is why you put `isExit()` in abstract Command class, kinda nice! 
I didn't use loop here, makes my life harder to use function to exit the recursive call 😮‍💨